pkg/status/combinedstatus-resolution.go

```diff
@@ -0,0 +1,308 @@
+		clauses, ok := statusCollectorNameToClauses[statusCollectorName]
+		if !ok {
+			delete(c.statusCollectorNameToData, statusCollectorName)
+			workStatusRemoved = true
+			continue
+		}
+
+		if !statusCollectorData.equals(clauses) {
+			c.statusCollectorNameToData[statusCollectorName].statusCollectorClauses = clauses
+			workStatusAdded = true
+		}
+	}
+
+	// add new statuscollector data and queue up all workstatuses. this does not qualify as
+	// a change in the resolution, since the workstatuses will update the statuscollector data
+	// when they are evaluated
+	for statusCollectorName, clauses := range statusCollectorNameToClauses {
+		if _, ok := c.statusCollectorNameToData[statusCollectorName]; !ok {
+			c.statusCollectorNameToData[statusCollectorName] = &statusCollectorData{
+				statusCollectorClauses: clauses,
+				workStatusRefToData:    make(map[util.ObjectIdentifier]workStatusData),
+			}
+
+			workStatusAdded = true
+		}
+	}
+
+	return workStatusRemoved, workStatusAdded
+}
+
+// getIdentifier returns the identifier of the combinedstatus resolution.
+// TODO: implement
+func (c *combinedStatusResolution) getIdentifier() util.ObjectIdentifier {
+	return util.ObjectIdentifier{}
+}
+
+// updateStatusCollector updates the status collector data in the
+// combinedstatus resolution. If the status collector is not relevant to the
+// latter, the function returns false. The function returns true if the status
+// collector data is updated. The given clauses are assumed to be valid and
+// immutable.
+func (c *combinedStatusResolution) updateStatusCollector(statusCollectorName string,
+	clauses *statusCollectorClauses) bool {
+	c.Lock()
+	defer c.Unlock()
+
+	scData, ok := c.statusCollectorNameToData[statusCollectorName]
+	if !ok {
+		return false // statusCollector is irrelevant to this combinedstatus resolution
+	}
+
+	if scData.equals(clauses) {
+		return false // statusCollector data is already up-to-date
+	}
+
+	// status collector clauses need to be updated, therefore update fields
+	// and invalidate all cached workstatus evaluations by resetting the map
+	c.statusCollectorNameToData[statusCollectorName] = &statusCollectorData{
+		statusCollectorClauses: clauses,
+		workStatusRefToData:    make(map[util.ObjectIdentifier]workStatusData),
+	}
+
+	return true
+}
+
+// evaluateWorkStatus evaluates the workstatus per relevant statuscollector.
+// The function returns true if any statuscollector data is updated.
+func (c *combinedStatusResolution) evaluateWorkStatus(ctx context.Context, celEvaluator *celEvaluator,
+	workStatusIdentifier util.ObjectIdentifier, workStatusContent *runtime.RawExtension) (bool, error) {
+	c.Lock()
+	defer c.Unlock()
+
+	logger := klog.FromContext(ctx)
+
+	updated := false
+	for statusCollectorName := range c.statusCollectorNameToData {
+		scData, ok := c.statusCollectorNameToData[statusCollectorName]
+		if !ok {
+			continue
+		}
```

Comment:
Why not just

```
for statusCollectorName, scData := range c.statusCollectorNameToData {
```

?

Reply:
I have to say I have no recollection of how this piece came to life. Again the code went through a change in implementation design, and I though publishing it before finalizing self-reviews to give the general idea (through the comments, mostly) would be beneficial. 

Though the latter is true regardless, I just apologize for having you check these instead.

Reply:
No worries. I understand that this is a work in progress. Appreciate the early view.